### ISSUE TITLE
docs build - process notes and TODOs

### ISSUE THREAD

**Issue by briantist** [ansible-collections/community.hashi_vault#138]
##### SUMMARY

This issue is meant to keep track of things related to the docs build process, which builds and publishes collection documentation on PRs and pushes, in temporary sites outside of the published collection docs on docs.ansible.com.

### On a PR (open, updated, resynced)
- The docs from the target are built (rendered)
- The docs (docsite changes only) from the PR are built
- Hashes of both are recorded to determine if the PR makes any differences
- A `git diff` of the builds is taken to show in markdown
- If there are changes, then:
  - The rendered docs are attached to the build as an artifact (expires in 7 days)
  - The rendered docs are published via https://surge.sh to a temporary site that's specific to the PR. This lets anyone browse the fully rendered docs by clicking a link.
  - A bot comment on the PR provides links to the rendered sites of both the target branch docs and the PR's docs for comparison, and shows the `git diff` output in an expandable section (unless it's too big, which crashes GHA), and it shows a list of modified files, each of which links to the published docsite.
- If there are no changes:
  - if there's an existing surge site (because there were previously changes in the PR), it is unpublished
  - if there's an existing PR comment (because of previous docs changes), it is removed.

In addition to the above, another build is triggered which does exactly the same thing, _but_ it will build the entire set of docs using the entire PR contents. That means **it will include changes to plugin and module documentation**.

The key for this one is that it gets built in a separate GitHub "environment" with rules that require it to be approved on every commit. This happens because the build is still privileged and we cannot trust the PR contents, so every commit must be reviewed by a person before approving this build. But once it runs, we get the complete picture of the built docs.

### On PR close or merge
- Any existing surge site is unpublished
- If there's an existing PR comment, it is updated
  - if the PR was closed, the comment is updated to say the site is unpublished
  - if the PR was merged, the comment is updated thanking for the contribution, and linking to the main surge docsite build (as the changes have now been incorporated there)

### On Push (to `main` only)
- the docs are built, and the "main" surge site is updated.

### The `pull_request_target` event
**REQUIRED READING:** https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

When a GitHub Workflow is run for the `pull_request` event, from a fork of the repository (the norm in this and most open source projects), the rights are limited, compared to PRs opened from a branch on the same repo (which requires write access to the repo). In particular, on a PR from a fork:
- the token used within workflow does not have write access to the repository, so it cannot (for example) commit, create/remove/update comments, open issues, etc.
- the workflow cannot read secrets

These are important security considerations, because anyone can open a PR, and use it to execute untrusted code, leading to unauthorized commits on the repo, secret exposure, etc.

For the purposes of this doc build process, this is important because:
- we need to store a secret for the token required to publish to surge
- we need to be able to created/edit/delete comments in PRs

#### Enter `pull_request_target`
Workflows run under this event _do_ have access to secrets, and _do_ have write access to the repository. The difference is that both the workflow definition itself, and its defined "ref" refer to the _target_ of the PR. That means that the workflow file is always loaded from the target branch (for example `main`) so that it cannot be manipulated in the PR, and it also means that using the `checkout` action with its defaults is going to clone the target (`ex `main`) and not the PR content.

The point of all this is: you can't trust the PR content when running in a context that has access to secrets and elevated privileges.

It _is_ possible to checkout the PR contents, but you must be careful not to execute anything in it, and there's lots of indirect ways that using the content of a PR could lead to exploitation, even without direct "execution".

#### Bringing it back to docs build
We use `pull_request_target` so we can get the surge secret and so we can post PR comments.

But in order to render the docs in the PR, we need some content from the PR.

In our case, we bring in only `docs/docsite/`, which contains YAML and RST. While a PR could stick other files there, we won't be executing anything in that subdirectory.

#### Problem 1: Changing the workflow
Because the workflow is always loaded from the target, it's very difficult to make changes to the workflow via the PR process and test them from within the same PR. The version of the workflow will always be loaded from `main`.

I've dealt with this in two ways:
##### Commit/merge it now, use other PRs, update incrementally
* For this, I'd put up a PR with a change to keep track of the purpose and what's being changed, and then merge it right away.
* A new PR would be put up that tests the functionality (it will be using the version of the workflow now in `main`)
* If anything needs to be changed in the workflow, I make new commits, put the original workflow PR # in the commit message (for example `fix type in docs workflow (#123)`), and push the commit to `main`. 
* By mentioning the PR, GitHub will automatically show it linked in the original PR, even though it was not part of the PR; that at least keeps some semblance of cohesiveness.

This isn't great; pushing commits directly to `main`, some of which are the type that squashing was _made_ for, and squashing those after the fact means a force push (also not good).

##### Put workflow changes on a direct branch, open other PRs against the branch
* In this method, the PR making changes to the workflow **must be in a branch in the actual repository, not in a fork**
* A new PR is still needed to test the changes, but this time, the PR should be made against the above branch, not against `main` (that is, you'll be asking to merge the PR into the branch that the workflow PR is based on). **This PR can be in a fork.**
* Updates to the workflow are pushed to the workflow PR branch as usual (just remember to push `upstream` if your `origin` is your fork)
* One difficulty here is that if you re-run a CI run on the test PR, it will not use the updated state of the workflow branch, it will use the one it ran against before (the exact commit). So if you update your workflow branch, your test PRs need some other change to trigger their CI to run against the new workflow. I've been using empty commits for that a la `git commit -m empty --allow-empty` and that's enough to re-trigger. 
* A test PR that doesn't contain actual changes you want will be closed anyway. A test PR with changes you might want to keep is going to be squashed, or a manual rebase is super easy (removing the empty commits altogether).

This method is a little more complex, but it avoids the direct pushes.

Both methods make it near impossible for someone without write access to the repo to meaningfully contribute to the docs build process.

#### Problem 2: which docs?
The docs build process includes plugin and module documentation, which is very nice. However because of the security implications of `pll_request_target`, we're only actually using the `docs/docsite/` folder from the PR. So changes made to collection content other than the docsite, will not be taken into account in the docs build.

The docstrings for executable content live in `.py` files, and even though some of them are read via AST and not executed, doc fragments _are_ executed, and as a result, I'm not sure that we can ever include them in PRs submitted on forks.

~One idea I had was to possibly introduce a `workflow_dispacth` event-based workflow (manual invocation) that can be run by a maintainer. This workflow would run against a PR, and render the entire docsite. The idea would be that a maintainer who has reviewed the PR can decide to run the docs build against the entire PR contents, and the resulting docs build site would contain all changes. A way of sorting gating that process. It would have to be re-done on every new commit to a PR as well.~

~Another idea I had was to figure out if the PR was submitted on a branch in the same repo, which can implicitly be trusted (the submitter had write access already). This doesn't help much on the external contributor front, but in many/most collections, the maintainers are very active contributors (it would certainly help me!).~

~Both of the above will make for some tricky logic and conditionals in the workflows, making them harder to read and grok.~

So this has been sorta-solved by the separate build that requires approval (scroll up).

#### Problem 3: workflows running _in_ forks
Right now, my workflow compares where it's running from to ensure it's running in `ansible-collections/community.hashi_vault`, and skips many/most steps if it's not. This was really an indirect way of saying "a fork doesn't have the surge secret so don't bother running all this stuff, it'll fail anyway".

It would be better to check for the existence of a secret instead, so that forks _could_ optionally add such a thing.

That leads to the possibility of using secrets to define things like the surge site name stub, etc.

It has some weird implications for PR comments though.

##### ISSUE TYPE
- Documentation Report

##### COMPONENT NAME

docs build

##### ANSIBLE VERSION

N/A

**Comment by webknjaz** [comment]
Any reason not to use RTD for PR builds? That part with surge feels like reinventing the wheel.

**Comment by briantist** [comment]
> Any reason not to use RTD for PR builds? That part with surge feels like reinventing the wheel.

RTD was the first thing I considered. But it doesn't seem like they have any kind of push mode. It appears to be git based where you connect it to github and it builds docs based on commits. Since I'm not committing rendered docs, and since I want to publish multiple simultaneous sets of docsites (that are specific to PRs), RTD seems to be a poor fit. I'm happy to hear otherwise but I could not find a way to make that work.

I also considered GitHub Pages, but there were several reasons that made that tricky as well. You can see more if you review the past DAWGs meetings minutes where this was discussed.

Surge was a surprisingly easy and straightforward option that came up in my search, and it's been working great for this purpose.

Of course anyone else who implements docs builds in their collections can choose to forego publishing, or choose a different publishing option.

**Comment by webknjaz** [comment]
Have you ever used RTD yourself? From your statements, it sounds like you haven't. It is able to build docs on pushes to branches on GitHub as well as to build separate docsites for each PR. Exactly like you want. I fail to see how you can see it as not solving the problem :man_shrugging:

For example, here's a site that gets updated on every commit to devel: https://ansible-pylibssh.rtfd.io. But each PR gets its own build.
If you look at https://github.com/ansible/pylibssh/pull/205, there's a status check reported from RTD, there's a link to a PR-specific website at https://ansible-pylibssh--205.org.readthedocs.build/en/205/ and while the build is in progress that link points to the build page at https://readthedocs.org/projects/ansible-pylibssh/builds/13759664/.

**Comment by webknjaz** [comment]
Are you saying that you want to run https://github.com/ansible-collections/community.hashi_vault/blob/main/docs/preview/build.sh on RTD but you don't know how? It's easy to work around (although, I'm truly confused why antsibull-doc isn't implemented as a proper Sphinx extension, but it's doable even without that).

**Comment by briantist** [comment]
> Have you ever used RTD yourself? From your statements, it sounds like you haven't. 

I definitely have not! I spent time researching it and was not able to figure out how to make it work for this use case. **Which is not to say that it isn't possible**, but their docs do not suggest that it is, and I was not able to find the information. I pursued other avenues, and found one that worked. 🤷‍♂️

> It is able to build docs on pushes to branches on GitHub as well as to build separate docsites for each PR. Exactly like you want. I fail to see how you can see it as not solving the problem 🤷‍♂️
> 
> For example, here's a site that gets updated on every commit to devel: https://ansible-pylibssh.rtfd.io. But each PR gets its own build.
> If you look at [ansible/pylibssh#205](https://github.com/ansible/pylibssh/pull/205), there's a status check reported from RTD, there's a link to a PR-specific website at https://ansible-pylibssh--205.org.readthedocs.build/en/205/ and while the build is in progress that link points to the build page at https://readthedocs.org/projects/ansible-pylibssh/builds/13759664/.

That looks great! Definitely what I was trying to do. From your links I still have no idea _how_ it was done; as I said I was able to get surge working and published in 3 minutes, far less time than it took me to even read the RTD docs, so I went with it.

---

Anyway, while I'm not at all opposed to switching to RTD, is there any reason you're so insistent upon it specifically?

**Comment by webknjaz** [comment]
The reason is that it's a de-facto standard in the wider Python community (outside of the Ansible bubble) and there's a good chance that more contributors are familiar with it. Besides, it supports Sphinx as a first-class citizen. This means that there would be no need to maintain a GHA+repo-specific custom setup (hence no need for people to gain this knowledge).
The problem you're facing is probably that you want to run arbitrary code before invoking Sphinx and RTD only runs dep installs before running sphinx. The solution is to move the `antsibull-doc` invocation into the Sphinx setup. It belongs there anyway. I asked you earlier about `build.sh` — I want to know if it needs to be executed only on RTD or in any env invoking sphinx?
Also, the PR builds have been introduced about 2 years ago, first as a private beta (that I participated) and then it became a default.
@briantist I need your RTD login (you can sign in via GH) so I'll be able to add you as a maintainer to a test project. Also, I need somebody with better repo access to click on a button that automatically sets up the webhooks.
I'll send a PR with the RTD config and edits to `conf.py` to show you how it works but we need somebody to enable the webhooks (otherwise, I can trigger manual builds only after the merge to the main branch).

**Comment by webknjaz** [comment]
@briantist I'm still waiting for your RTD account + some answers. I'm on PTO now so expect to get a PR from me after Sep 29.

**Comment by briantist** [comment]
@webknjaz the `setup.sh` and the entire Sphinx config all come from the work @felixfontein has done with `antsibull-docs sphinx-init` in https://github.com/ansible-community/antsibull/pull/297 . 

If you feel that should be changed or done in a different place, I encourage you to propose and/or implement those changes within that tool, as I don't intend on modifying that output all at much. I treat them mostly as generated files (I modified `build.sh` slightly to not use hardcoded paths).

What you're proposing sounds great and I look forward to it but it might be best worked on with other members of the community; I know Felix has been interested in working on docs build and we were going to work on getting some of my changes into the collections he manages. Maybe it'd be best to skip that and for you to work with him on how best to do it via RTD from the get go?

Also strongly encouraging you to join and attend the [Documentation Working Group](https://github.com/ansible/community/issues/579), as your input would have been valuable while all of this was being worked on.

For the short term I'm not intending to switch to RTD because I have something "good enough" and need to focus efforts elsewhere for a little while, but I'll be watching closely to see how that plays out and I hope to use it in the future! Thanks!

**Comment by webknjaz** [comment]
> @webknjaz the `setup.sh` and the entire Sphinx config all come from the work @felixfontein has done with `antsibull-docs sphinx-init` in [ansible-community/antsibull#297](https://github.com/ansible-community/antsibull/pull/297) .

Did you mean `build.sh`?

> If you feel that should be changed or done in a different place, I encourage you to propose and/or implement those changes within that tool, as I don't intend on modifying that output all at much. I treat them mostly as generated files (I modified `build.sh` slightly to not use hardcoded paths).
> 
> What you're proposing sounds great and I look forward to it but it might be best worked on with other members of the community; I know Felix has been interested in working on docs build and we were going to work on getting some of my changes into the collections he manages. Maybe it'd be best to skip that and for you to work with him on how best to do it via RTD from the get go?

I still want to send a demo PR here because you cannot demo anything on that repo. It doesn't have to be merged but it'd be a good platform for discussions, a more visual one.
And then, it could be moved to antsibull then.

> Also strongly encouraging you to join and attend the [Documentation Working Group](https://github.com/ansible/community/issues/579), as your input would have been valuable while all of this was being worked on.

I wish I had time for this. Maybe I'll join sometimes but often I'm just unable to come 🤷‍♂️

**Comment by briantist** [comment]
> Did you mean `build.sh`?

Yes :)

> I still want to send a demo PR here because you cannot demo anything on that repo. It doesn't have to be merged but it'd be a good platform for discussions, a more visual one.
> And then, it could be moved to antsibull then.

Sure, that's ok as long as it doesn't require me signing up for RTD and sending the credentials, it's just something I'm not able to put much time to right now. I suggested working with Felix on one of the collections he maintains as there are a lot of choices, and I know he already has at least one RTD account, and a sample collection build RTD site ( https://github.com/felixfontein/felix-ansible-docsite ), so may be more fruitful than it would be working with me for the time being (if he's available).

> > Also strongly encouraging you to join and attend the [Documentation Working Group](https://github.com/ansible/community/issues/579), as your input would have been valuable while all of this was being worked on.
> 
> I wish I had time for this. Maybe I'll join sometimes but often I'm just unable to come 🤷‍♂️

Of course, I can definitely relate, time is a precious commodity for us all, and I do appreciate you spending some of yours on reviews and suggestions!

**Comment by felixfontein** [comment]
> I know he already has at least one RTD account, and a sample collection build RTD site ( https://github.com/felixfontein/felix-ansible-docsite ),

I don't have an RTD account. That site I'm building locally and uploading on my own server.

**Comment by briantist** [comment]
> > I know he already has at least one RTD account, and a sample collection build RTD site ( https://github.com/felixfontein/felix-ansible-docsite ),
> 
> I don't have an RTD account. That site I'm building locally and uploading on my own server.

My mistake! Apologies for the assumption.

**Comment by webknjaz** [comment]
FWIW you can log in to RTD via GitHub with one click and that's about what you'll need to do 🤷‍♂️

**Comment by briantist** [comment]
In #199, I've run into an issue with the split docsite-only/vs. full build:
- a new docsite page references two new plugins that are added in the same PR
- the limited docsite build then emits warnings about the references, since they can't be resolved (the plugins are not included in this buid), which fails the workflow
- the full docs build works fine

A "quick" resolution I am considering is:
- the limited docs build failure should not count toward the PR
- if I can use exit codes to determine whether "failure" was caused only by warnings vs errors, I could continue past the step on warnings and let the limited docs build workflow finish (capturing warnings to show in the PR comment would be nice)

As a longer term fix, I am looking at using [the `permissions:` key in the workflow](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#example-1-passing-the-github_token-as-an-input) to possibly get rid of the split build altogether. The way I envision this working is as follows:
- the docs build itself would run in a separate job (same workflow), where the elevated permissions granted by the `pull_request_target` event would be purposefully dropped.
- That would resolve the concern about executing arbitrary PR code.
- The PR author cannot change the workflow that is executed to get around that.
- This job would upload the rendered docsite as an artifact, and would also preserve the diff output that's used in the PR comment (will have to see if this can be done as an output value or will also need to be put into an artifact).
- A second job (same workflow), depending on the first job, will use its elevated permissions to:
  - publish the site using the surge secret
  - post the PR comment

This ought to get rid of the need for the second job that requires manual approval on every push, which I have to also pay attention to to make sure it doesn't run before limited build (otherwise it will clobber the full build). So this would be better all around.

One thing I'm not sure of with regard to dropping permission, is whether I can actually drop the permission to read secrets. I don't see that as one of the possible scopes listed by GitHub, so I will need to do some experimenting. If I can get this working, it would be a huge step up for this.

**Comment by briantist** [comment]
docs build has moved to a dedicated repository!
https://github.com/ansible-community/github-docs-build/

Further discussion, ideas, and documentation will happen there, in the issues, discussions, and wiki in that repo.
